python/lsst/daf/butler/datastores/fileDatastore.py

```diff
@@ -413,11 +423,27 @@ def _get_stored_records_associated_with_refs(
             The matching records indexed by the ref ID.  The number of entries
             in the dict can be smaller than the number of requested refs.
         """
-        records = self._table.fetch(dataset_id=[ref.id for ref in refs])
+        # Check datastore records in refs first.
+        records_by_ref: defaultdict[DatasetId, list[StoredFileInfo]] = defaultdict(list)
+        refs_with_no_records = []
+        for ref in refs:
+            if ref.datastore_records is None:
+                refs_with_no_records.append(ref)
+            else:
+                if (ref_records := ref.datastore_records.get(self._table.name)) is not None:
+                    # Need to make sure they have correct type.
+                    for ref_record in ref_records:
+                        if not isinstance(ref_record, StoredFileInfo):
+                            raise TypeError(
+                                f"Datastore record has unexpected type {ref_record.__class__.__name__}"
+                            )
+                        records_by_ref[ref.id].append(ref_record)
+
+        # If there were any refs without datastore records, check opaque table.
+        records = self._table.fetch(dataset_id=[ref.id for ref in refs_with_no_records])
```

Comment:
This possibility is something I'd imagined disappearing from `Datastore` on DMTN-249: I was hoping we could make sure all possibly-relevant records were passed to a datastore whenever we'd get to code like this, such that if any were missing it'd represent repo corruption or logic bugs and hence the datastore should raise.  And that'd mean it wouldn't have a `self._table` at all.

Do you see this as transitional code, or something that'd stay for the long haul?

Reply:
This is indeed transitional, the final goal is to remove `self._table` completely. I just keep it because not all functionality is replaced yet and there are many unit tests to be updated.